Feature: raw clipboard access

livecodefraser opened the pull request:
This PR adds a set of new accessors for the clipboard and drag-and-drop clipboard.

At the lowest level, "rawClipboardData" and "rawDragData" properties have been added. These provide low-level access to the system clipboard and are platform-specific knowledge is required in order to use them effectively.

Built on top of that are the "fullClipboardData" and "fullDragData" properties. These provide richer access to the clipboard than is available using the "clipboardData" and "dragData" properties (which, as of this PR, are deprecated).

This has resulted in completely re-writing the MCPasteboard code. The old implementation as been removed. Now, the platform-specific elements are confined to the MCRawClipboard implementations. The MCClipboard class provides most of the behaviour of the "fullClipboardData" and "fullDragData" properties.

Care has been taken to preserve the behaviour of the existing "clipboardData" and "dragData" properties. Because this is legacy behaviour, this behaviour is implemented in the exec methods for those properties rather than by complicating MCClipboard with conflicting behaviours.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/livecode/livecode/2946)

<!-- Reviewable:end -->

livecodefraser committed:
Add a mixin class providing basic reference-counting semantics

livecodefraser committed:
[[ RawClipboard ]] Initial in-engine API for raw clipboard access

livecodefraser committed:
Add a const-variant of MCAutoArray::operator[]

livecodefraser committed:
[[ RawClipboard ]] Initial implementation for OSX

livecodefraser committed:
[[ RawClipboard ]] Add script hooks for the raw clipboard (getters only)

livecodefraser committed:
[[ RawClipboard ]] Add clipboard locking and implement setting

livecodefraser committed:
[[ RawClipboard ]] Update OSX implementation to use locking and enable setting

livecodefraser committed:
[[ RawClipboard ]] Code style fixes

livecodefraser committed:
Add a non-retaining constructor for MCAutoValueRef classes

This constructor is only usable via explicit constructor syntax
(i.e `MCAutoValueRef t_value(p_value)` syntax).

livecodefraser committed:
Remove some of the header insanity

object.h now includes the headers is requires rather than having
to add them in each place it is included.

livecodefraser committed:
Const-ify various MCAutoRefcounted accessors

livecodefraser committed:
Refactor a couple of methods in MCField taking MCStrings

livecodefraser committed:
[[ RawClipboard ]] Rename the MCClipboard syntax class

This has been renamed to MCClipboardProp to avoid clashes with
the new MCClipboard class used to provide the in-engine clipboard
interface.

livecodefraser committed:
[[ RawClipboard ]] Refactor the existing clipboard/drag code

All the clipboard and drag code now goes through MCRawClipboard.

livecodefraser committed:
[[ RawClipboard ]] OSX platform changes for the refactored clipboard code

livecodefraser committed:
[[ RawClipboard ]] Remove the old clipboard code (generic and OSX for now)

livecodefraser committed:
[[ RawClipboard ]] OSX: only push changes if updates have been made

livecodefraser committed:
[[ RawClipboard ]] Keep the high-level and raw clipboards in sync

livecodefraser committed:
[[ RawClipboard ]] OSX: update the list of known type <-> UTI mappings

livecodefraser committed:
[[ RawClipboard ]] OSX: Add special-casing for "public.file-url" clipboard data

livecodefraser committed:
[[ RawClipboard ]] Refactor MCClipboard::AddText slightly

livecodefraser committed:
[[ RawClipboard ]] Reject modifications to externally-provided clipboard data

livecodefraser committed:
[[ RawClipboard ]] Fix some minor typos

livecodefraser committed:
[[ RawClipboard ]] OSX: don't pull changes if clipboard is unchanged

livecodefraser committed:
[[ RawClipboard ]] OSX: AddRepresentation should replace existing reps

livecodefraser committed:
[[ RawClipboard ]] Implement getting and setting of file lists

livecodefraser committed:
Add a method to allow re-assigning MCAutoValueRefs

livecodefraser committed:
[[ RawClipboard ]] Don't overwrite user 'text' when adding 'files'

livecodefraser committed:
[[ RawClipboard ]] Fix a Win32 compilation error

livecodefraser committed:
[[ RawClipboard ]] Change the API for encoding file lists

livecodefraser committed:
Add MCAutoDataRefArray

livecodefraser committed:
[[ RawClipboard ]] Fix a pointer cast in the refactored field code

livecodefraser committed:
[[ RawClipboard ]] Implementation for Win32

livecodefraser committed:
[[ RawClipboard ]] Fix mac server building

livecodefraser committed:
Fix the Linux Poll implementation to also poll GLib's FDs

Previously, these weren't getting included when polling so some
UI events were not being delivered in a timely manner.

livecodefraser committed:
[[ RawClipboard ]] Fix various refcounting bugs

livecodefraser committed:
Add a non-retaining setter to MCAutoValueRef

livecodefraser committed:
[[ RawClipboard ]] Implement for Linux

livecodefraser committed:
[[ RawClipboard ]] Linux: add multiple formats for each type

This is required as various apps appear to look for different
strings for each type (for example, there are 3 different MIME
types used to represent RTF).

livecodefraser committed:
[[ RawClipboard ]] Linux: wait for response before getting clipboard data

The X11 clipboard model is asynchronous so we need to wait for
the clipboard owner to send us the data before attempting to read
it.

livecodefraser committed:
[[ RawClipboard ]] Linux: work around a bug when pre-allocating MCDataRef space

livecodefraser committed:
[[ RawClipboard ]] Linux: support some special clipboard types

livecodefraser committed:
[[ RawClipboard ]] Linux: make sure we return the TARGETS list with the correct type

livecodefraser committed:
[[ RawClipboard ]] Linux: fix issues with the drag-and-drop loop

livecodefraser committed:
[[ RawClipboard ]] Update the middle-click clipboard before pasting

livecodefraser committed:
[[ RawClipboard ]] Linux: fix incompatibilities with external drag-and-drop sources

livecodefraser committed:
[[ RawClipboard ]] Fix an off-by-1 error when pasting files to the OSX clipboard

livecodefraser committed:
[[ RawClipboard ]] Linux: robustness fixes when putting files on the clipboard

livecodefraser committed:
[[ RawClipboard ]] Don't clear the selection clipboard when a field gains focus

livecodefraser committed:
[[ RawClipboard ]] OSX: fix handling of file paths

livecodefraser committed:
[[ RawClipboard ]] Windows: remove an un-needed class

livecodefraser committed:
[[ RawClipboard ]] Windows: fix dragging into LiveCode from other programs

livecodefraser committed:
[[ RawClipboard ]] Windows: support IStream data objects

livecodefraser committed:
[[ RawClipboard ]] Ensure that 'the keys of the clipboarddata' doesn't contain duplicates

livecodefraser committed:
Merge branch 'develop' of https://github.com/runrev/livecode into feature-raw_clipboard

Conflicts:
	engine/engine-sources.gypi
	engine/src/dispatch.cpp
	engine/src/dsklnx.cpp
	engine/src/executionerrors.h
	engine/src/lnxdce.cpp
	engine/src/lnxdclnx.cpp
	engine/src/lnxdnd.cpp
	engine/src/lnxtransfer.cpp
	engine/src/object.h
	engine/src/parsedef.h
	engine/src/platform.cpp
	engine/src/transfer.h
	libfoundation/include/foundation-auto.h

livecodefraser committed:
[[ RawClipboard ]] Add documentation for 'rawClipboardData' and 'lock/unlock clipboard'

peter-b commented:
@livecodefraser Please include a description for this pull request, to facilitate featuring it in TWiL. Thank you! :+1:

runrevmark commented:
@trevordevore: No - the clipboardData is unchanged functionality wise. If you want to add multiple formats you have to use the rawClipboardData. The higher-level clipboard feature does need reworking at some point to make it more flexible, but that has backwards-compability issues. @livecodefraser has done a good job here though - the high-level clipboard implementation now sits on top of the platform-specific rawClipboardData implementation, meaning it should be easier to revisit the higher-level support at a later date much more easily.

runrevmark commented:
@trevordevore: I wonder if 'private' should actually be sideline to everything? i.e. so you can set both private and another key. Perhaps file an enhancement request? I'm not sure whether that would have any impact on current semantics...

runrevmark commented:
@trevordevore: Okay - so having had some chats with @livecodefraser - as it turns out there's more features here than original posed. We are retaining complete backwards compatibility with existing clipboard functionality via 'the clipboard' function and 'the clipboardData'. They will still choose a single format and present it and its related formats.

We're going to add a 'fullClipboardData' property which basically gives the kind of clipboard access you really need without having to resort to the rawClipboardData. It will present named keys for common formats (text, rtf, html, gif, png, jpeg, tiff); then synthesized/engine-specific formats on top (plaintext, styledtext, htmltext, rtftext, image, files, objects, private). The image key will derive from one of the image types. The *text keys will derive from the most representative text type (styles, rtf, text) and pass through the engine's conversion functions.

Most importantly, fullClipboardData will give you multiple formats to get and set. (You'll either be able to set the engine-level ones, and it will synthesize the low-level ones appropriately as it does now; or you'll be able to set the lower-level ones and the engine will synthesize the appropriate high-level ones).

trevordevore commented:
@runrevmark - that sounds fantastic and solves a real need. Thanks @livecodefraser!

livecodefraser committed:
[[ RawClipboard ]] Add syntax for rawDragData, full{Clipboard,Drag}Data

The syntax is currently unimplemented and results in linking errors
as the exec methods aren't implemented yet.

livecodefraser committed:
[[ RawClipboard ]] Restore old behaviour for the clipboardData property

livecodefraser committed:
[[ RawClipboard ]] Fix a NULL pointer crash in MCClipboard::GetLegacyOrdering()

livecodefraser committed:
[[ RawClipboard ]] Implement support for 'the fullClipboardData' and 'the fullDragData'

livecodefraser commented:
@trevordevore - I've now added a global "the fullClipboardData" property which allows you to put multiple types of data on the clipboard. Not written up the documentation yet, but it needs to be inside a lock/unlock clipboard pair. It'll probably also need to be cleared before writing to it, by doing "set the fullClipboardData to empty".

There are also equivalents to all of the clipboard enhancements for "the dragData" - "the rawDragData" and "the fullDragData". Unlike the clipboard, no locking is required for these as the drag events related to drag-and-drop imply the locking.

livecodefraser committed:
Fix a missing semicolon in lnxdce.cpp

trevordevore commented:
@livecodefraser now we are cooking with gas!

livecodefraser committed:
[[ RawClipboard ]] Add docs for fullClipboardData and fullDragData

livecodefraser committed:
Merge branch 'develop' of https://github.com/runrev/livecode into feature-raw_clipboard

Conflicts:
	engine/src/fieldf.cpp
	engine/src/globals.h
	engine/src/util.h

livecodefraser committed:
[[ RawClipboard ]] Add a release note

livecodefraser commented:
@trevordevore I've now added some documentation on the new clipboard accessors. If you have time, I'd love any feedback you have on whether this provides the clipboard functionality that you need. Now's the time to get it right, before we commit to it!

trevordevore commented:
@livecodefraser I just checked out your branch and will review it later today.

FourthWorld commented:
@livecodefraser  Impressive work, Fraser.  Thanks,

trevordevore commented:
@livecodefraser I've been playing around with the fullclipboarddata. I was able to successfully put text and images on the clipboard and paste them into other apps! Very cool.
1. Once I've copied something to the clipboard using the fullClipboardData, I can no longer paste text on the clipboard into SE. Maybe an issue with clipoardData? Once I triggered the bug I tried <put the clipboarddata> in the message box Xcode caught an exception when calling MCNameIsEqualTo(p_key, MCN_text).
2. I would suggest modifying the example in the docs in the example for fullclipboarddata and fulldragdata. The example giver the impression that you have to set the htmltext and text separately. The engine will fill in the text if you set the htmltext.
3. If an error occurs while the clipboard is locked and you aren't able to unlock the clipboard then you get error the next time you try to lock the clipboard (clipboard is locked). Do we need a lockClipboard (like lockscreen) property so that we only lock the clipboard if it isn't already locked? Or should the clipboard automatically unlock itself like lock screen and lock messages?
4. put the keys of fullclipboardData complains that I need to lock the clipboard. Could the lock happen automatically in that case? Is there any harm or lost functionality in doing so?
5. I get an error when trying to access the fullClipboardData["stylesArray"]. clipboard: bad item type or data.

runrevmark commented:
@trevordevore: Sounds like we need to do a bit more work on the locking front :)

The clipboard is notionally a shared object among processes. We decided to make locking it explicit to access fullClipboardData and rawClipboardData to ensure consistency.

If you do:

```
if "foo" is among the keys of the fullClipboardData then
  ... doSomethingWith the fullClipboardData["foo"]
end if
```

Then it isn't guaranteed that "foo" will still be among the keys of the fullClipboardData as another process might have modified it in the interim. Basically, any clipboard operation at a low-level requires the current process to lock it in some fashion.

We did consider tying the lifetime of the access to something like 'until the next event processing' however I was concerned about that because it makes things fragile:

```
if "foo" ... then
  ... someThirdPartyLibraryHandlerOrOperationWhichCallsWaitButYouDontKnow...
  ... doSomethingWIth ... "foo"
end if
```

trevordevore commented:
@runrevmark: Makes sense. If locking is the price we have to pay for full clipboard access then I'm willing to pay it :-)

trevordevore commented:
@runrevmark: Question though - should <the clipboard> be marked as deprecated as it has been in the new docs? For many purposes it is easier to use and sufficient.

runrevmark commented:
@trevordevore: Well, I think we will have to relax the locking requirement somehow as the same problem affects clipboardData too in fact (more so with full because you might want to access two formats) (the clipboardData does implicitly lock / unlock). If we can find a reasonable lock lifetime rule which is easy to see and use, then having implicit locking would be better. I wonder if implicit locking until the end of the current handler _might_ make sense; if you don't explicitly lock, that is (in which case the lock should probably reset at return to the root event loop).

In terms of deprecation of some existing syntax - that is a good question - we'll reconsider that.

Simplicity and correctness are in direct contention here!

livecodefraser committed:
[[ RawClipboard ]] Remove deprecation notices from clipboard and clipboardData docs

Changes to the design of the new clipboard features mean that they
still provide useful functionality, after all.

livecodefraser committed:
[[ RawClipboard ]] Explicitly clear the clipboard when copying field contents, an image or LiveCode objects

livecodefraser commented:
@trevordevore looks like the in-engine copies (field contents, images, objects) were failing to clear the clipboard beforehand, so they got merged with the existing data. Because the existing data was earlier in the list of data representations, it would tend to get pasted instead of the new data. I couldn't reproduce it locally -- I suspect it depends quite strongly on what type of data has already been placed and where you're pasting it.

The deprecation notes have also been removed from clipboard and clipboardData. I think our design was a little bit different when that was originally decided.

Your comment on the example is worth taking into account. I already know how it works so finding a good example is a bit tricky. It also suggests that I need to clarify that setting any one of text,htmltext,rtftext,styles,styledArray will implicitly generate the others too.

The locking is very strict at the moment and I'm not really happy with how it works. As Mark said, we've not come up with a workable alternative yet, though.

The 'stylesArray' issue is probably that I've forgotten to hook something up. I'm taking a look at that now. I added it as the infrastructure was mostly there and it has non-overlapping uses compared to 'styles' - the array form can be modified while the serialised form can be passed to other LiveCode instances.

trevordevore commented:
@livecodefraser I tested your last commits and can confirm that I can copy and paste within the SE after I've modified the fullClipboardData. So it appears that is fixed.

I'm still having trouble pasting the fullClipboardData into the SE. If I execute the following code, trying to paste into the SE doesn't work.

```
on mouseUp   
   lock clipboard

   set the fullClipboardData to empty
   set the fullClipboardData["image"] to the text of image 1
   set the fullClipboardData["htmltext"] to "<p>What is <i>your</i> quest?</p>"
   set the fullClipboardData["private"] to "I'm private!"

   unlock clipboard
end mouseUp
```

I just checked `the clipboard` and it returns "image". So it appears that the issue is just one of the clipboard looking for images before text. The IDE just needs to be updated to use `the fullClipboardData` so that it pastes when text is on the clipboard, regardless of what else is on the clipboard.

Also, I still get a crash if I try to access `the clipboardData` after setting `the fullClipboardData`. I don't if I specify a key such as `the clipboardData["text"]`.

runrevmark commented:
@livecodefraser: I wonder if 'stylesArray' should actually be 'styledText' - to give synergy with the field property (like htmlText, rtfText etc.)?

livecodefraser committed:
[[ RawClipboard ]] Implement the getting of the stylesArray property

livecodefraser committed:
[[ RawClipboard ]] Rename "stylesArray" to "styledText"

livecodefraser commented:
We're getting closer, I hope!

fullClipboardData["stylesArray"] is now working, except that it is now called fullClipboardData["styledText"], as Mark suggested.

Is the pasting failure exclusive to the SE or are you seeing it in all fields? The default pasting behaviour for fields should now favour text, if it is available.

trevordevore commented:
@livecodefraser I see it in all fields. `the clipboard` returns 'image' so the engine is definitely giving preference to the image on the clipboard as opposed to the text.

livecodefraser committed:
[[ RawClipboard ]] Fix a crash when reading clipboardData with no subscript

livecodefraser commented:
From the looks of it, the preference for images is coming from the IDE. If the IDE wasn't intercepting the paste request and doing it itself, the in-engine implementation would see that a field is focused and would paste as text instead of as an image.

Unfortunately, we can't change the fact that clipboardData prefers to return "image" to "text" as that would break existing stacks that want to paste images. I'm going to prod @runrevmark for an opinion on whether changing the IDE's behaviour is workable (we could update it to either check the fullClipboardData and choose type based on context or just stop intercepting the events and let the engine handle it).

runrevmark commented:
@livecodefraser: We will definitely want to update the IDE appropriately... If the engine now does the correct thing with pasting text or images in the right context, then we can just remove the IDE code. Failing that, maybe ask @livecodeali to take a look and see what updating the IDE would do?

FourthWorld commented:
@livecodefraser  RE "Unfortunately, we can't change the fact that clipboardData prefers to return "image" to "text" as that would break existing stacks that want to paste images." - who expects that?

Perhaps I'm overlooking some use case that may be useful to someone else, but if the clipboardData contains both image and text data AND I have an insertion point in an editable field I've always expected that the textual data would get priority.  If there is not editable field open, or if there is no textual data in the clip, only then would I expect an image to be pasted.

I'd be happy to run this past the use-list to see if anyone is expecting anything else, but it would seem like an odd user experience if they were.

Besides, if someone really wants an odd user experience, they could still get one by working with the clipboardData directly, while the paste command would do the most logical thing for the context at hand, no?

livecodefraser commented:
The engine does exactly what you're asking for already :smile: . The problem is the IDE - it intercepts the paste request and, instead of letting the engine do it, it performs the paste. Because it does this by looking at the clipboardData, it doesn't know that there are both images and text so you end up with a new image object being pasted.

This is actually rather unfortunate behaviour for another reason too: it means that pasting behaves differently in standalones.

In summary, my opinion is that the engine is right and the IDE is wrong. Hopefully nobody will object if we change the IDE's paste behaviour, but the clipboardData has to stay as-is (the paste command is not implemented on top of the clipboardData so this doesn't pose a problem).

FourthWorld commented:
Thanks @livecodefraser, well said: "In summary, my opinion is that the engine is right and the IDE is wrong."  So like the old joke goes, "Then don't do that". :)

If we make the engine smart enough to let people build apps easily, it should be possible to write simpler IDE code as well.

Know the engine.
Trust the engine.
Use the engine.

livecodeali commented:
As far as I can see, most of the IDE's interfering with pasting is for validation - in almost all cases to prevent multi-line pasting into single-line fields. These I guess will still need to be there in some form or other...

FourthWorld commented:
Shouldn't the developer be in control of how much text goes into a field of a given size?

livecodeali commented:
In IDE stacks, not user stacks. For example, the single line message box (although this is implemented incorrectly at the moment, causing this http://quality.runrev.com/show_bug.cgi?id=16018)

trevordevore commented:
@livecodeali This brings up an interesting point. I rarely use the paste command in an app because I usually want to cleanse any data being pasted in. Whether it is removing newlines in a single line field or stripping out unnecessary formatting there is often something that needs to be done. Is there some sort of help that the engine could provide when pasting that would allow a developer more control? Maybe a message that is sent that would allow the user to massage the data and then whatever is returned the engine pastes in? Just typing out loud here.

```
on PasteClipboard pClipboardKey, pContent
   if pClipboardKey is "text" then
      -- go through and remove newlines from pContent
   end if

   return pContent
end PasteClipboard
```

FourthWorld commented:
Thanks for the clarification, @livecodeali . Would it simplify things or complicate them to have Edit -> Paste respond by using the built-in paste command for all user stacks, and for IDE stacks it would dispatch a paste message, which could then be handled within the IDE stack as appropriate (or perhaps in one backscript if the desired behavior is universal among IDE stacks).

livecodeali commented:
@FourthWorld If we had something like @trevordevore 's idea or something like it we could just pass the message if if the target is not an IDE stack, allowing the engine's paste.

One of the problems we have at the moment, which has only just this second become clear to me, is that loads of places in the IDE only handle pasting by handling the commandKeyDown message. Hence, in prior versions of the IDE, the paste from the menu would do the paste by dispatching commandKeyDown appropriately to the target object.

It would be better if any method of pasting could be validated properly rather than converting them all to a key shortcut.

FourthWorld commented:
I like messages in general, but my concern here is that if we have a paste message one could arguably expect a copy message as well, and then we're off into the weeds wondering how many other messages make sense once that paradigm is established, and then we need to consider the implications....

"Real artists ship."

If it were up to me I'd suggest we  simplify the IDE paste handling using what we have in hand today, and continue this discussion in the forums to flesh out engine-level enhancements for the future.

trevordevore commented:
@FourthWorld @livecodefraser @livecodeali I just did a test and the behavior of `the clipboard` pre-fullClipboardData and post-fullClipboardData is different. If `the clipboard` worked how it use to then the IDE would continue to work as it currently does. The IDE should still be updated at some point, but you could defer it.

Here is the test I did:

Put multiple types of data on the clipboard using @livecodefraser 's branch:

```
on mouseUp   
   lock clipboard

   set the fullClipboardData to empty
   set the fullClipboardData["image"] to the text of image 1
   set the fullClipboardData["htmltext"] to "<p>What is <i>your</i> quest?</p>"
   set the fullClipboardData["private"] to "I'm private!"

   unlock clipboard
end mouseUp
```

Check `the clipboard` and it returns "image".

Now switch over to LC 8 or LC 7.1 and `the clipboard` returns "text", even though the system clipboard contains the same data. Shouldn't `the clipboard` continue to return "text" for backwards compatibility?

In LC 7.1

livecodefraser commented:
@trevordevore Hmm, for that particular series of insertions, the old clipboard should be returning "image" - because it was added first, it should be taking priority over the later "htmltext" that was added.

runrevmark commented:
@FourthWorld: Clearly the appropriate thing here is to ensure that the 'old' syntax works the way as it did before (which _should_ be possible) and then the IDE will continue to work - which I suspect is precisely what @livecodefraser has already been investigating.

However, there is always value in exploring ways to improve things generally and in the future, is there not?

Whilst 'messages for actions' would not be conflated with clipboard improvements (they are distinct things) - there is a 'gem' of an idea forming in the above - although it is not quite there yet.

There are several things at play here:
- There is a menu item 'paste'.
- There is an engine command 'paste'
- There is the observation that targets of 'paste' often want to customize what is pasted, or do it themselves.

Now, one can replace 'paste' with any imperative action that the user might want to perform in a user interface (these are generally menuitems, but can also be keyboard shortcuts). The process should probably be this:
- I can mark a menu item as invoking an action.
- An action is a named thing which causes a message to be sent to the currently focused element.
- The message can be either handled or passed.
- If passed, and if it reaches the engine, the engine performs the default action for the message.

This is basically how bindings between menu items / keyboard shortcuts and actions work in Cocoa for all intents and purposes. Cocoa has a responder chain which is akin to our message path. This abstraction means that you wouldn't have to have code in menuitems, and you wouldn't have to have target specific code in your menuitem dispatch function. Instead, you can localize handling of an action at the level which is most appropriate.

There is definitely a general mechanism lurking here (which will be needed to allow widgets to easily interact with common actions such as paste without scripting - like other engine controls do) - it perhaps needs some more thought at a later date.

livecodefraser commented:
Looking at the old code, it actually looks like a bug in our old OSX implementation - both Linux and Windows will take an image over text if they find it first but Mac will always take text, regardless of ordering... though more reading reveals another bug: it looks like it doesn't calculate priorities correctly, so text will always win out over images, regardless of platform. Sigh.

Given the conflict between the apparent intention of the code and the actual results of the code, I'm not sure which behaviour we should adopt. There is a choice between "correctness" (highest priority item wins) or "compatibility" (do the old, accidental, un-documented thing). Compatibility usually wins in LiveCode but it is what led to me marking it as deprecated in the first place - it doesn't do the right thing so will lead to strange bugs later on (if you put an image with some alt-text on the clipboard, for example, you lose the ability to paste that image using the clipboardData).

trevordevore commented:
@runrevmark menus invoking actions? Yes please.

@livecodefraser Yes, I have had bug report open on this very issue for quite some time.

http://quality.runrev.com/show_bug.cgi?id=10637

I think it should be fixed as you can't get at images placed on the clipboard from Safari. I guess that does bring us back to getting the IDE to paste properly if the cursor is in a field and both text and an image are on the clipboard.

livecodefraser committed:
[[ RawClipboard ]] Fix a potential NULL dereference

livecodefraser committed:
[[ RawClipboard ]] Allow implicit locking and clearing for the fullClipboardData

livecodefraser committed:
[[ RawClipboard ]] Update the docs to reflect the implicit locking for fullClipboardData

Explicit locking is still highly recommended, of course.

livecodefraser committed:
[[ RawClipboard ]] Restore the buggy clipboardData behaviour

Due to a bug in the old code, images were never chosen when text
was also available. This bug has existed for long enough to become
ossified as a feature so this patch restores that behaviour.

The code that implements the behaviour originally intended is
still present but disabled, just in case we ever want to switch
to using it.

livecodefraser commented:
I have now restored the previous behaviour of clipboardData: text will always be chosen in preference to images (this should fix the IDE pasting bug).

Because I had already implemented it, I've left the 'corrected' method available, should we decide that this decision needs to be re-visited. I'm going to have to mark bug 10637 as unfixable for now and I'll add a note pointing to this discussion. I'll also point out the new fullClipboardData property as a work-around.

livecodefraser committed:
[[ RawClipboard ]] Linux: don't talk to the system clipboard in server builds

livecodefraser committed:
[[ RawClipboard ]] Add the ability to set the fullClipboardData to an array

runrevmark commented:
Reviewed 2 of 86 files at r1, 1 of 12 files at r2, 1 of 17 files at r4.
Review status: 4 of 90 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[engine/src/clipboard.cpp, line 114 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QTRQRl0Z-6QEeIWdJ)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/clipboard.cpp#L114)):</sup>
What is this method for?

<!-- Reviewable comment -K0QTRQSa832JANvcvan:-279830286 -->

---

<sup>**[engine/src/control.cpp, line 1629 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QUTfYfWYLiKBMLG2n)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/control.cpp#L1629)):</sup>
Why the need for checking two methods here? Surely if the dragboard 'HasTextOrCompatible()' then it means it could convert text from styledtext?

<!-- Reviewable comment -K0QUTfYfWYLiKBMLG2o:-873899500 -->

---

<sup>**[engine/src/desktop-dc.cpp, line 1071 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QUn2dApwvH8RuS0ZY)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/desktop-dc.cpp#L1071)):</sup>
How is the PlatformDragDrop code now interacting with the pasteboard?

<!-- Reviewable comment -K0QUn2dApwvH8RuS0ZZ:-1143953606 -->

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/2946)

<!-- Sent from Reviewable.io -->

runrevmark commented:
Reviewed 4 of 86 files at r1.
Review status: 8 of 90 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

<sup>**[engine/src/newobj.cpp, line 391 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QXctj7a71nhGPnbUJ)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/newobj.cpp#L391)):</sup>
Niggly I know - but this should probably be MCClipboardFunc...

<!-- Reviewable comment -K0QXctksVfm9h9N0p-T:1099805170 -->

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/2946)

<!-- Sent from Reviewable.io -->

runrevmark commented:
Reviewed 34 of 86 files at r1, 4 of 12 files at r2, 7 of 17 files at r4, 3 of 5 files at r5.
Review status: 56 of 90 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

<sup>**[engine/src/exec-pasteboard.cpp, line 399 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QbdlotIsviLEX2FYm)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/exec-pasteboard.cpp#L399)):</sup>
This should be an MCN_styled_text 'constant'  - along with the ones below also :)

<!-- Reviewable comment -K0QbdlpgfraKErUPWUx:-1950404375 -->

---

<sup>**[engine/src/executionerrors.h, line 2673 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0Qc63xJM5-CjDsjBke)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/executionerrors.h#L2673)):</sup>
I'm pretty sure all error constants are of the form EE_<module>_<error> where <error> contains no '_'.

<!-- Reviewable comment -K0Qc63xJM5-CjDsjBkf:186177611 -->

---

<sup>**[engine/src/field.cpp, line 1257 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QcHQFWFyucsOe77MT)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/field.cpp#L1257)):</sup>
Is there a reason not to do this now?

<!-- Reviewable comment -K0QcHQFWFyucsOe77MU:1377817982 -->

---

<sup>**[engine/src/globals.cpp, line 508 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0Qcmnng0Lz7sylvHLT)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/globals.cpp#L508)):</sup>
Doesn't the clipboard lock need to be recursive? If so, how is this used?

<!-- Reviewable comment -K0QcmnoAigU16DUIGPt:-1910921635 -->

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/2946)

<!-- Sent from Reviewable.io -->

runrevmark commented:
Reviewed 11 of 86 files at r1, 1 of 1 files at r3, 1 of 17 files at r4.
Review status: 69 of 90 files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

<sup>**[docs/dictionary/command/lock-clipboard.lcdoc, line 21 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QqOXx_-93ddAukMEM)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/docs/dictionary/command/lock-clipboard.lcdoc#L21)):</sup>
This isn't quite true anymore is it? We now have implicit locking semantics for all clipboard stuff?

<!-- Reviewable comment -K0QqOXzT-W-pVERq5qi:-1047644161 -->

---

<sup>**[docs/dictionary/command/lock-clipboard.lcdoc, line 23 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QqYku0LjXak6VNsXJ)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/docs/dictionary/command/lock-clipboard.lcdoc#L23)):</sup>
Mention fullClipboardData too?

<!-- Reviewable comment -K0QqYku0LjXak6VNsXK:550540661 -->

---

<sup>**[docs/dictionary/command/lock-clipboard.lcdoc, line 30 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QqaMxNsMs-636bHiE)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/docs/dictionary/command/lock-clipboard.lcdoc#L30)):</sup>
Add fullClipboardData

<!-- Reviewable comment -K0QqaMxNsMs-636bHiF:-626677171 -->

---

<sup>**[docs/dictionary/command/unlock-clipboard.lcdoc, line 1 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QqdjYfnI_lsQt8ZNm)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/docs/dictionary/command/unlock-clipboard.lcdoc#L1)):</sup>
Same considerations as lock clipboard

<!-- Reviewable comment -K0QqdjYfnI_lsQt8ZNn:-2126275525 -->

---

<sup>**[docs/dictionary/property/clipboardData.lcdoc, line 72 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QqoWFRhZufhimNIGX)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/docs/dictionary/property/clipboardData.lcdoc#L72)):</sup>
Need to mention fullClipboardData and explain why it is better!

<!-- Reviewable comment -K0QqoWFRhZufhimNIGY:-522139039 -->

---

<sup>**[docs/dictionary/property/fullClipboardData.lcdoc, line 35 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QqtNxMe1t4UbfhTlm)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/docs/dictionary/property/fullClipboardData.lcdoc#L35)):</sup>
Move this up to be next to the other text keys, move styles down to be next to objects

<!-- Reviewable comment -K0QqtNxMe1t4UbfhTln:817189684 -->

---

<sup>**[docs/dictionary/property/fullClipboardData.lcdoc, line 46 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0Qr1VF7gGVYgbF04Sa)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/docs/dictionary/property/fullClipboardData.lcdoc#L46)):</sup>
Might be better as a 'Note' - more keys and data types might be added in the future (not 'exhaustive' suggests we have missed describing some, rather than we might add more in the future).

<!-- Reviewable comment -K0Qr1VF7gGVYgbF04Sb:-245632858 -->

---

<sup>**[docs/dictionary/property/fullClipboardData.lcdoc, line 61 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QrEln_ahs0jFtNU-n)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/docs/dictionary/property/fullClipboardData.lcdoc#L61)):</sup>
I'd suggest making this entry the one which explains all the keys. So we need to mention what 'private' is :)

<!-- Reviewable comment -K0QrEln_ahs0jFtNU-o:-1234345015 -->

---

<sup>**[docs/dictionary/property/fullDragData.lcdoc, line 57 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QrLTgP8ojP2iMApCF)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/docs/dictionary/property/fullDragData.lcdoc#L57)):</sup>
Same considerations as fullClipboardData.

<!-- Reviewable comment -K0QrLTgP8ojP2iMApCG:2012274839 -->

---

<sup>**[docs/dictionary/property/rawClipboardData.lcdoc, line 29 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QrPIOumDTB48kxzXP)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/docs/dictionary/property/rawClipboardData.lcdoc#L29)):</sup>
Is this true under the new implicit locking?

<!-- Reviewable comment -K0QrPIOumDTB48kxzXQ:-566976565 -->

---

<sup>**[docs/dictionary/property/rawClipboardData.lcdoc, line 48 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QrUfQ_h6DfRYp94ki)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/docs/dictionary/property/rawClipboardData.lcdoc#L48)):</sup>
I'm not sure this is correct - clearing the clipboard is equivalent to taking ownership. Perhaps we should use that metaphor?

<!-- Reviewable comment -K0QrUfQ_h6DfRYp94kj:1332813260 -->

---

<sup>**[engine/src/lnx-clipboard.cpp, line 221 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0Qs8JCE_aTwXUOPhdM)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/lnx-clipboard.cpp#L221)):</sup>
This seems a little overzealous - perhaps an assert instead? After all you've split string, which by definition gives you an array of strings...

<!-- Reviewable comment -K0Qs8JDffevtJQGRH_I:229477300 -->

---

<sup>**[engine/src/lnx-clipboard.cpp, line 284 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QsR0JFOj1m2FBrVF-)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/lnx-clipboard.cpp#L284)):</sup>
Again - this would be better done as a static assertion I think.

<!-- Reviewable comment -K0QsR0JFOj1m2FBrVF0:2134477740 -->

---

<sup>**[engine/src/lnx-clipboard.cpp, line 289 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QsTrBXuijRFaYs5J5)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/lnx-clipboard.cpp#L289)):</sup>
We're trying to get rid of these sorts of cast... Perhaps make t_path a stringref?

<!-- Reviewable comment -K0QsTrBXuijRFaYs5J6:-710563220 -->

---

<sup>**[engine/src/lnxdnd.cpp, line 45 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QtO78-1d42BaB9JMd)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/lnxdnd.cpp#L45)):</sup>
If we are going to use Finalize, we should probably use Initialize ;)

<!-- Reviewable comment -K0QtO78-1d42BaB9JMe:-880494643 -->

---

<sup>**[engine/src/lnxdnd.h, line 1 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0Qta5gAej4erGlFzTa)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/lnxdnd.h#L1)):</sup>
Can we ditch this file altogether?

<!-- Reviewable comment -K0Qta5gAej4erGlFzTb:369874959 -->

---

<sup>**[engine/src/mac-clipboard.mm, line 264 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QtvRSWs2R6BgjkGEe)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/mac-clipboard.mm#L264)):</sup>
Perhaps '7' should be replaced by 'URL_FILE_PREFIX_LENGTH', and 'file://' by 'URL_FILE_PREFIX_CSTRING'?

<!-- Reviewable comment -K0QtvRTa0ddaF2MlsiG:-1680577431 -->

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/2946)

<!-- Sent from Reviewable.io -->

livecodefraser committed:
Merge branch 'develop' of https://github.com/runrev/livecode into feature-raw_clipboard

livecodefraser committed:
[[ RawClipboard ]] Remove the now-empty lnxdnd.h header

livecodefraser committed:
[[ RawClipboard ]] Rename the clipboard execution error constants

They have been changed to be more consistent with the existing
error code names.

livecodefraser committed:
[[ RawClipboard ]] Rename class MCClipboardProp to MCClipboardFunc

It's a function, not a property, after all!

livecodefraser committed:
[[ RawClipboard ]] Rename the MCLinuxDragAndDropInit function

It is now MCLinuxDragAndDropInitialize, for consistency.

livecodefraser committed:
[[ RawClipboard ]] Change some run-time checks to assertions

They were sanity checking something that shouldn't ever happen in
the first place.

livecodefraser commented:
Review status: 69 of 90 files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

<sup>**[engine/src/clipboard.cpp, line 114 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QTRQRl0Z-6QEeIWdJ)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/clipboard.cpp#L114)):</sup>
It isn't implemented yet ;)

When a drag-and-drop enters a LiveCode window, we grab the incoming data and (depending on platform) either cache the metadata or all of the data. Once the drag leaves a LiveCode window, this method gets called with the intention of dropping the caches, otherwise the data will be retained until the next drag-and-drop operation happens.

<!-- Reviewable comment -K0Vbc66yq0lAu6pUewb:-76346119 -->

---

<sup>**[engine/src/field.cpp, line 1257 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QcHQFWFyucsOe77MT)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/field.cpp#L1257)):</sup>
Because I can't figure out which of the eleventy different ways of adding text to a field is the right one here! Any pointers would be appreciated.

<!-- Reviewable comment -K0VcHH6f2D0e6g9-_Xn:2136989992 -->

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/2946)

<!-- Sent from Reviewable.io -->

livecodefraser committed:
[[ RawClipboard ]] Remove a redundant check in field drag-and-drop checking

MCClipboard::HasTextOrCompatible() includes all formats supported
by HasStyledTextOrCompatible().

livecodefraser committed:
[[ RawClipboard ]] Use MCN_* constants for clipboard types

runrevmark commented:
Review status: 69 of 90 files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

<sup>**[engine/src/field.cpp, line 1257 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0QcHQFWFyucsOe77MT)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/engine/src/field.cpp#L1257)):</sup>
Middle click is the same as 'paste' but for the 'transient' selection... Follow the method used in MCDispatch::dopaste when MCactivefield != nil :)

<!-- Reviewable comment -K0VsbfLVwIWfpR_eUbU:-540081454 -->

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/2946)

<!-- Sent from Reviewable.io -->

livecodefraser committed:
[[ RawClipboard ]] Make the clipboard lock recursive

livecodefraser committed:
Prevent one of the MCAutoValueRef constructors from retaining nil

livecodefraser committed:
[[ RawClipboard ]] Fix some missing semicolons in the Linux clipboard code

livecodefraser committed:
[[ RawClipboard ]] Add a comment explaining why middle-click doesn't paste styled text

Basically, it causes more problems than it solves.

livecodefraser committed:
[[ RawClipboard ]] Implement the MCClipboard::ReleaseData method

This method clears the clipboard contents if they came from an
external source. It is used to clear left-over data from an
external drag-and-drop operation.

livecodefraser committed:
[[ RawClipboard ]] Various fixes to the new clipboard documentation

livecodefraser committed:
[[ RawClipboard ]] Add some tests for clipboardData and fullClipboardData

Not all of these pass yet...

livecodefraser committed:
[[ RawClipboard ]] Only run the clipboard tests on desktop platforms

livecodefraser committed:
Fix the detection of JPEG images when adding to the clipboard

runrevmark commented:
Reviewed 1 of 86 files at r1, 2 of 17 files at r4, 1 of 5 files at r5, 28 of 32 files at r10.
Review status: 93 of 97 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

<sup>**[docs/dictionary/command/unlock-clipboard.lcdoc, line 28 [r9]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0aqCdtOGY5rTOQKQk_)** ([raw file](https://github.com/livecode/livecode/blob/54d0a49ed60b4d588fa35038ac93a74d832954fd/docs/dictionary/command/unlock-clipboard.lcdoc#L28)):</sup>
fullClipboardData :)

<!-- Reviewable comment -K0aqCduHgYSTFtW_tMz:1566746147 -->

---

<sup>**[docs/dictionary/property/fullClipboardData.lcdoc, line 59 [r10]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0awxjCbEUQqvgv_Swe)** ([raw file](https://github.com/livecode/livecode/blob/36fd3f8ffe69024da1565b1930a3cea91e8928a5/docs/dictionary/property/fullClipboardData.lcdoc#L59)):</sup>
This should be fillClipboardData (not fullDragData)

<!-- Reviewable comment -K0awxjCbEUQqvgv_Swf:-1434575465 -->

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/2946)

<!-- Sent from Reviewable.io -->

runrevmark commented:
@livecode-fraser: Almost there - just a few niggles with the docs :)

livecodefraser committed:
[[ RawClipboard ]] More corrections to the clipboard docs

livecodefraser committed:
Merge branch 'develop' of https://github.com/runrev/livecode into feature-raw_clipboard

Conflicts:
	engine/src/executionerrors.h

runrevmark commented:
Reviewed 2 of 19 files at r11.
Review status: 79 of 97 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

<sup>**[docs/dictionary/property/fullClipboardData.lcdoc, line 59 [r10]](https://reviewable.io:443/reviews/livecode/livecode/2946#-K0awxjCbEUQqvgv_Swe)** ([raw file](https://github.com/livecode/livecode/blob/36fd3f8ffe69024da1565b1930a3cea91e8928a5/docs/dictionary/property/fullClipboardData.lcdoc#L59)):</sup>
Sorry - I mean - should the fullDragData be there?

<!-- Reviewable comment -K0ebmyXqL56y1E-gz1A:802040517 -->

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/2946)

<!-- Sent from Reviewable.io -->

runrevmark commented:
Reviewed 12 of 19 files at r11.
Review status: 91 of 97 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/2946)

<!-- Sent from Reviewable.io -->

runrevmark commented:
Reviewed 4 of 19 files at r11.
Review status: 95 of 97 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/2946)

<!-- Sent from Reviewable.io -->

livecodefraser committed:
Add a missing include to dllst.h

livecodefraser committed:
[[ RawClipboard ]] Auto-convert text to styled text when added

livecodefraser committed:
[[ RawClipboard ]] Linux: don't use GDK in no-UI mode

runrevmark commented:
Reviewed 1 of 32 files at r10, 1 of 19 files at r11, 5 of 5 files at r12.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/2946)

<!-- Sent from Reviewable.io -->

runrevmark commented:
@livecode-vulcan review ok fe3607c

runrevmark commented:
@livecode-vulcan review ok dbe7c61

livecode-vulcan commented:
💙 review by @runrevmark ok dbe7c61ffc6d69f57a7b51d1520b055b79740f11

livecode-vulcan commented:
💥 merge conflict dbe7c61ffc6d69f57a7b51d1520b055b79740f11 → 007858f392c8826b94d698221f0820720628e159

livecodefraser committed:
Merge branch 'develop' of https://github.com/runrev/livecode into feature-raw_clipboard

Conflicts:
	engine/src/executionerrors.h

runrevmark commented:
@livecode-vulcan review ok 5d07862

livecode-vulcan commented:
💙 review by @runrevmark ok 5d07862a40eb3a17a4c5842c4a64013c6762f8bb

livecode-vulcan commented:
😞 test failure 5d07862a40eb3a17a4c5842c4a64013c6762f8bb
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/760)
- try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/774)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/761)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/763)
- try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/774)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/784)
- try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/776)

livecodefraser committed:
[[ RawClipboard ]] Add a dummy implementation for Android

livecodefraser committed:
[[ RawClipboard ]] Add a dummy implementation for iOS

runrevmark commented:
@livecode-vulcan review ok d46d976

livecode-vulcan commented:
💙 review by @runrevmark ok d46d976da1eb3b99b6faee8b8ce10c23602566cd

livecode-vulcan commented:
😞 test failure d46d976da1eb3b99b6faee8b8ce10c23602566cd
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/770)
- try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/784)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/771)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/773)
- try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/784)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/794)
- try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/786)

livecodefraser committed:
[[ RawClipboard ]] Add the dummy iOS and Android files to Gyp

runrevmark commented:
@livecode-vulcan review ok f0118f2

livecode-vulcan commented:
💙 review by @runrevmark ok f0118f2dace095679e2e580da4ec398fa86fe877

livecode-vulcan commented:
😎 test success f0118f2dace095679e2e580da4ec398fa86fe877
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/797)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/811)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/798)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/801)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/811)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/821)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/814)
